File: src/php/WorldCountries.php
<?php

namespace DevMe\WorldCountries;

class WorldCountries
{
    private array $countries;

    public function __construct()
    {
        $jsonPath = __DIR__ . '/../../data/world-countries.json';
        $jsonContent = file_get_contents($jsonPath);
        $this->countries = json_decode($jsonContent, true);
    }

    public function getAll(): array
    {
        return $this->countries;
    }

    public function getByAlpha2Code(string $code): ?array
    {
        $code = strtoupper($code);
        foreach ($this->countries as $country) {
            if ($country['cca2'] === $code) {
                return $country;
            }
        }
        return null;
    }

    public function getByAlpha3Code(string $code): ?array
    {
        $code = strtoupper($code);
        foreach ($this->countries as $country) {
            if ($country['cca3'] === $code) {
                return $country;
            }
        }
        return null;
    }

    public function getByNumericCode(string $code): ?array
    {
        foreach ($this->countries as $country) {
            if ($country['ccn3'] === $code) {
                return $country;
            }
        }
        return null;
    }

    public function getByCommonName(string $name): ?array
    {
        $nameLower = strtolower($name);
        foreach ($this->countries as $country) {
            if (strtolower($country['name']['common']) === $nameLower) {
                return $country;
            }
        }
        return null;
    }

    public function getByOfficialName(string $name): ?array
    {
        $nameLower = strtolower($name);
        foreach ($this->countries as $country) {
            if (strtolower($country['name']['official']) === $nameLower) {
                return $country;
            }
        }
        return null;
    }

    public function getByRegion(string $region): array
    {
        $result = [];
        $regionLower = strtolower($region);
        foreach ($this->countries as $country) {
            if (strtolower($country['region']) === $regionLower) {
                $result[] = $country;
            }
        }
        return $result;
    }

    public function getBySubregion(string $subregion): array
    {
        $result = [];
        $subregionLower = strtolower($subregion);
        foreach ($this->countries as $country) {
            if (strtolower($country['subregion']) === $subregionLower) {
                $result[] = $country;
            }
        }
        return $result;
    }

    public function getByCurrency(string $currencyCode): array
    {
        $result = [];
        $code = strtoupper($currencyCode);
        foreach ($this->countries as $country) {
            if (isset($country['currencies'][$code])) {
                $result[] = $country;
            }
        }
        return $result;
    }

    public function getByLanguage(string $languageCode): array
    {
        $result = [];
        $code = strtolower($languageCode);
        foreach ($this->countries as $country) {
            if (isset($country['languages'][$code])) {
                $result[] = $country;
            }
        }
        return $result;
    }

    public function getByCapital(string $capital): array
    {
        $result = [];
        $capitalLower = strtolower($capital);
        foreach ($this->countries as $country) {
            foreach ($country['capital'] as $cap) {
                if (strtolower($cap) === $capitalLower) {
                    $result[] = $country;
                    break;
                }
            }
        }
        return $result;
    }

    public function getByContinent(string $continent): array
    {
        $result = [];
        $continentLower = strtolower($continent);
        foreach ($this->countries as $country) {
            foreach ($country['continents'] as $cont) {
                if (strtolower($cont) === $continentLower) {
                    $result[] = $country;
                    break;
                }
            }
        }
        return $result;
    }

    public function getByBorders(string $countryCode): array
    {
        $result = [];
        $code = strtoupper($countryCode);
        foreach ($this->countries as $country) {
            if (in_array($code, $country['borders'])) {
                $result[] = $country;
            }
        }
        return $result;
    }

    public function getLandlocked(): array
    {
        $result = [];
        foreach ($this->countries as $country) {
            if ($country['landlocked'] === true) {
                $result[] = $country;
            }
        }
        return $result;
    }

    public function getIndependent(): array
    {
        $result = [];
        foreach ($this->countries as $country) {
            if ($country['independent'] === true) {
                $result[] = $country;
            }
        }
        return $result;
    }

    public function getUNMembers(): array
    {
        $result = [];
        foreach ($this->countries as $country) {
            if ($country['unMember'] === true) {
                $result[] = $country;
            }
        }
        return $result;
    }

    public function search(string $query): array
    {
        $result = [];
        $searchTerm = strtolower($query);
        
        foreach ($this->countries as $country) {
            $found = false;
            
            if (stripos($country['name']['common'], $searchTerm) !== false ||
                stripos($country['name']['official'], $searchTerm) !== false ||
                stripos($country['cca2'], $searchTerm) !== false ||
                stripos($country['cca3'], $searchTerm) !== false) {
                $found = true;
            }
            
            if (!$found) {
                foreach ($country['capital'] as $capital) {
                    if (stripos($capital, $searchTerm) !== false) {
                        $found = true;
                        break;
                    }
                }
            }
            
            if (!$found) {
                foreach ($country['altSpellings'] as $spelling) {
                    if (stripos($spelling, $searchTerm) !== false) {
                        $found = true;
                        break;
                    }
                }
            }
            
            if ($found) {
                $result[] = $country;
            }
        }
        
        return $result;
    }

    public function getByTimezone(string $timezone): array
    {
        $result = [];
        foreach ($this->countries as $country) {
            if (in_array($timezone, $country['timezones'])) {
                $result[] = $country;
            }
        }
        return $result;
    }

    public function getByCallingCode(string $code): array
    {
        $result = [];
        foreach ($this->countries as $country) {
            $fullCode = $country['idd']['root'] . implode('', $country['idd']['suffixes']);
            if (strpos($fullCode, $code) !== false) {
                $result[] = $country;
            }
        }
        return $result;
    }

    public function getAllCurrencies(): array
    {
        $currencies = [];
        foreach ($this->countries as $country) {
            foreach (array_keys($country['currencies']) as $currency) {
                $currencies[] = $currency;
            }
        }
        return array_unique($currencies);
    }

    public function getAllLanguages(): array
    {
        $languages = [];
        foreach ($this->countries as $country) {
            foreach (array_keys($country['languages']) as $language) {
                $languages[] = $language;
            }
        }
        return array_unique($languages);
    }

    public function getAllCapitals(): array
    {
        $capitals = [];
        foreach ($this->countries as $country) {
            foreach ($country['capital'] as $capital) {
                $capitals[] = $capital;
            }
        }
        return array_unique($capitals);
    }

    public function getAllRegions(): array
    {
        $regions = [];
        foreach ($this->countries as $country) {
            $regions[] = $country['region'];
        }
        return array_unique($regions);
    }

    public function getAllSubregions(): array
    {
        $subregions = [];
        foreach ($this->countries as $country) {
            $subregions[] = $country['subregion'];
        }
        return array_unique($subregions);
    }

    public function getAllContinents(): array
    {
        $continents = [];
        foreach ($this->countries as $country) {
            foreach ($country['continents'] as $continent) {
                $continents[] = $continent;
            }
        }
        return array_unique($continents);
    }

    public function getAllTimezones(): array
    {
        $timezones = [];
        foreach ($this->countries as $country) {
            foreach ($country['timezones'] as $timezone) {
                $timezones[] = $timezone;
            }
        }
        return array_unique($timezones);
    }

    public function count(): int
    {
        return count($this->countries);
    }

    public function getByArea(?float $minArea = null, ?float $maxArea = null): array
    {
        $result = [];
        foreach ($this->countries as $country) {
            if ($minArea !== null && $country['area'] < $minArea) {
                continue;
            }
            if ($maxArea !== null && $country['area'] > $maxArea) {
                continue;
            }
            $result[] = $country;
        }
        return $result;
    }

    public function getByPopulation(?int $minPop = null, ?int $maxPop = null): array
    {
        $result = [];
        foreach ($this->countries as $country) {
            if ($minPop !== null && $country['population'] < $minPop) {
                continue;
            }
            if ($maxPop !== null && $country['population'] > $maxPop) {
                continue;
            }
            $result[] = $country;
        }
        return $result;
    }

    public function sortByName(string $order = 'asc'): array
    {
        $countries = $this->countries;
        usort($countries, function($a, $b) use ($order) {
            $result = strcmp($a['name']['common'], $b['name']['common']);
            return $order === 'desc' ? -$result : $result;
        });
        return $countries;
    }

    public function sortByPopulation(string $order = 'desc'): array
    {
        $countries = $this->countries;
        usort($countries, function($a, $b) use ($order) {
            $result = $a['population'] - $b['population'];
            return $order === 'desc' ? -$result : $result;
        });
        return $countries;
    }

    public function sortByArea(string $order = 'desc'): array
    {
        $countries = $this->countries;
        usort($countries, function($a, $b) use ($order) {
            $result = $a['area'] - $b['area'];
            return $order === 'desc' ? -$result : $result;
        });
        return $countries;
    }
}

File: src/ruby/world_countries.rb
require 'json'

module WorldCountries
  class Countries
    attr_reader :countries

    def initialize
      json_path = File.join(File.dirname(__FILE__), '..', '..', 'data', 'world-countries.json')
      json_content = File.read(json_path)
      @countries = JSON.parse(json_content)
    end

    def all
      @countries
    end

    def get_by_alpha2_code(code)
      code_upper = code.upcase
      @countries.find { |country| country['cca2'] == code_upper }
    end

    def get_by_alpha3_code(code)
      code_upper = code.upcase
      @countries.find { |country| country['cca3'] == code_upper }
    end

    def get_by_numeric_code(code)
      @countries.find { |country| country['ccn3'] == code.to_s }
    end

    def get_by_common_name(name)
      name_lower = name.downcase
      @countries.find { |country| country['name']['common'].downcase == name_lower }
    end

    def get_by_official_name(name)
      name_lower = name.downcase
      @countries.find { |country| country['name']['official'].downcase == name_lower }
    end

    def get_by_region(region)
      region_lower = region.downcase
      @countries.select { |country| country['region'].downcase == region_lower }
    end

    def get_by_subregion(subregion)
      subregion_lower = subregion.downcase
      @countries.select { |country| country['subregion'].downcase == subregion_lower }
    end

    def get_by_currency(currency_code)
      code_upper = currency_code.upcase
      @countries.select { |country| country['currencies'].key?(code_upper) }
    end

    def get_by_language(language_code)
      code_lower = language_code.downcase
      @countries.select { |country| country['languages'].key?(code_lower) }
    end

    def get_by_capital(capital)
      capital_lower = capital.downcase
      @countries.select do |country|
        country['capital'].any? { |cap| cap.downcase == capital_lower }
      end
    end

    def get_by_continent(continent)
      continent_lower = continent.downcase
      @countries.select do |country|
        country['continents'].any? { |cont| cont.downcase == continent_lower }
      end
    end

    def get_by_borders(country_code)
      code_upper = country_code.upcase
      @countries.select { |country| country['borders'].include?(code_upper) }
    end

    def get_landlocked
      @countries.select { |country| country['landlocked'] == true }
    end

    def get_independent
      @countries.select { |country| country['independent'] == true }
    end

    def get_un_members
      @countries.select { |country| country['unMember'] == true }
    end

    def search(query)
      search_term = query.downcase
      @countries.select do |country|
        country['name']['common'].downcase.include?(search_term) ||
        country['name']['official'].downcase.include?(search_term) ||
        country['cca2'].downcase.include?(search_term) ||
        country['cca3'].downcase.include?(search_term) ||
        country['capital'].any? { |cap| cap.downcase.include?(search_term) } ||
        country['altSpellings'].any? { |alt| alt.downcase.include?(search_term) }
      end
    end

    def get_by_timezone(timezone)
      @countries.select { |country| country['timezones'].include?(timezone) }
    end

    def get_by_calling_code(code)
      @countries.select do |country|
        full_code = country['idd']['root'] + country['idd']['suffixes'].join
        full_code.include?(code.to_s)
      end
    end

    def all_currencies
      currencies = []
      @countries.each do |country|
        currencies.concat(country['currencies'].keys)
      end
      currencies.uniq.sort
    end

    def all_languages
      languages = []
      @countries.each do |country|
        languages.concat(country['languages'].keys)
      end
      languages.uniq.sort
    end

    def all_capitals
      capitals = []
      @countries.each do |country|
        capitals.concat(country['capital'])
      end
      capitals.uniq.sort
    end

    def all_regions
      @countries.map { |country| country['region'] }.uniq.sort
    end

    def all_subregions
      @countries.map { |country| country['subregion'] }.uniq.sort
    end

    def all_continents
      continents = []
      @countries.each do |country|
        continents.concat(country['continents'])
      end
      continents.uniq.sort
    end

    def all_timezones
      timezones = []
      @countries.each do |country|
        timezones.concat(country['timezones'])
      end
      timezones.uniq.sort
    end

    def count
      @countries.size
    end

    def get_by_area(min_area: nil, max_area: nil)
      @countries.select do |country|
        next false if min_area && country['area'] < min_area
        next false if max_area && country['area'] > max_area
        true
      end
    end

    def get_by_population(min_pop: nil, max_pop: nil)
      @countries.select do |country|
        next false if min_pop && country['population'] < min_pop
        next false if max_pop && country['population'] > max_pop
        true
      end
    end

    def sort_by_name(order: 'asc')
      sorted = @countries.sort_by { |country| country['name']['common'] }
      order == 'desc' ? sorted.reverse : sorted
    end

    def sort_by_population(order: 'desc')
      sorted = @countries.sort_by { |country| country['population'] }
      order == 'desc' ? sorted.reverse : sorted
    end

    def sort_by_area(order: 'desc')
      sorted = @countries.sort_by { |country| country['area'] }
      order == 'desc' ? sorted.reverse : sorted
    end
  end
end